xprof/tally.cpp

```diff
@@ -1,350 +1,227 @@
-                                      (process_id_t)0), // process
-                      std::make_tuple(2, &bt_field_integer_unsigned_get_value,
-                                      (thread_id_t)0), // threads
-                      std::make_tuple(4, &bt_field_integer_signed_get_value,
-                                      (int)0)); // backend
-
-  const bt_field *common_context_field =
-      bt_event_borrow_common_context_field_const(event);
+//! User data collection structure.
+//! It is used to collect interval messages data, once data is collected. 
+//! It is aggregated and tabulated for printing.
+//! This structure can hold data that the user needs to be present among different 
+//! callbacks calls, for instance command line params that can modify the printing 
+//! behaviour.
+struct tally_dispatch_s {
+    //! User params provided to the user component.
+    btx_params_t *params;
+
+    //! Maps "level" with the names of the backends that appeared when processing host messages (lttng:host). 
+    //! This information is split by level. Refer to the "backend_level" array at the top of this 
+    //! file to see which backends may appear on each level. 
+    std::map<unsigned,std::set<const char*>> host_backend_name;
+
+    //! Maps "level" with the duration data collected from host messages (lttng:host) for every (host,pid,tid,api_call_name) entity.
+    //! EXAMPLE: map{ 0 => umap{ tuple("iris",1287,2780,"ompt_target") => TallyCoreTime } }
+    std::map<unsigned,std::unordered_map<hpt_function_name_t, TallyCoreTime>> host;
+
+    //! Maps "level" with the duration data collected from device messages (lttng:device) for every (host,pid,tid,api_call_name) entity.
+    //! EXAMPLE: map{ 2 => umap{ tuple("iris",1287,2780,"zeMemoryCopy") => TallyCoreTime } }
+    std::unordered_map<hpt_device_function_name_t, TallyCoreTime> device;
+
+    //! Maps "level" with the names of the backends appearing when processing traffic messages (lttng:traffic). 
+    //! This information is split by level. Refer to the "backend_level" array at the top of this 
+    //! file to see which backends may appear on each level. 
+    std::map<unsigned,std::set<const char*>> traffic_backend_name;
+
+    //! Maps "level" with the duration data collected from traffic messages (lttng:traffic) for every (host,pid,tid,api_call_name) entity.
+    //! EXAMPLE: map{ 0 => umap{ tuple("iris",1287,2780,"ompt_target") => TallyCoreTime } }
+    std::map<unsigned,std::unordered_map<hpt_function_name_t, TallyCoreByte>> traffic;
+
+    //! Maps a "(host,pid,tid,device_id)" with the device name.
+    //! The device name is collected when processing data of "device_name" messages (lttng:device_name). 
+    //! This assume that a process is attached to a device (with a given name), once the program execution starts, 
+    //! and this will not change during the execution of the program.
+    std::unordered_map<hp_device_t, std::string> device_name;
```

Comment:
I think we can put the first `unsigned` correspond to the `level` (it's non obvious, maybe we should put an ew type to make it clearer) and then merge all the comments. It's obvious what traffic, device, host mean IMO